docs/2.10.0/docs/upgrade/smart-contract-upgrades.rst

```diff
@@ -0,0 +1,1946 @@
+   be explicitly used when resolving package-name *ambiguities* in
+   either command template-id or interface resolution.
+
+   -  **Note:** The Command’s
+      :ref:`package_id_selection_preference <com.daml.ledger.api.v1.Commands.package_id_selection_preference>`
+      must not lead to ambiguous resolutions for package-names,
+      meaning that it must not contain two package-ids pertaining to
+      the same package-name otherwise the submission will fail with
+      an ``INVALID_ARGUMENT`` error
+
+Dynamic package resolution in Ledger API queries
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+When subscribing for
+:ref:`transaction <transaction-trees>`
+or :ref:`active contract streams <active-contract-service>`,
+users can now use the ``#<package-name>`` selector in the template-id format
+to specify that they’re interested in fetching events for all templates
+pertaining to the specified package-name. This template selection set is
+dynamic and it widens with each uploaded template/package.
+
+**Note:** The by-package-name query mechanism described here does not
+apply to events sourced from non-upgradable templates (coming from
+packages with LF < 1.16)
+
+Example
+^^^^^^^
+
+Given the following packages with LF 1.16 existing on the participant
+node:
+
+-  Package AppV1
+
+   -  package-name: ``app1``
+
+   -  package-version: ``1.0.0``
+
+   -  template-ids: ``pkgId1:mod:T``
+
+-  Package AppV2
+
+   -  package-name: ``app1``
+
+   -  package-version: ``2.0.0``
+
+   -  template-ids: ``pkgId2:mod:T``
+
+If a transaction query is created with a templateId specified as
+``#app1:mod:T``, then the events stream will include events from both
+template-ids: ``pkgId1:mod:T`` and ``pkgId2:mod:T``
+
+Codegen
+-------
+
+For packages that support SCU (i.e. LF1.16), generated code will use
+package-names in place of package-ids in template ids. Retrieved data
+from the ledger will be subject to the upgrade transformations described
+in previous sections.
+
+Concretely, this is implemented as follows:
+
+Java
+~~~~
+
+The classes that are generated for each template and interface contain a
+``TEMPLATE_ID`` field, which, for upgradable packages, will now use a
+package name, rather than a package id. In order to allow determining
+the package id of these packages, we add a new ``PACKAGE_ID`` field to all
+such classes. Upgradable packages will also cause ``PACKAGE_NAME`` and
+``PACKAGE_VERSION`` fields to be present.
+
+TypeScript
+~~~~~~~~~~
+
+The ``templateId`` field on generated template classes has been updated to
+use the package-name as the package qualifier for upgrade compatible
+packages. This will be used for command submission and queries, however
+note that the package qualifier given back in queries will contain the
+package-id, rather than package-name. Generated modules will now also
+give the package “reference”, which will be the package-name for upgrade
+compatible packages, otherwise it’ll be the package-id.
```

Comment:
```suggestion
packages. This is used for command submission and queries. However,
note that the package qualifier given back in queries contains the
package-id, rather than the package-name. Generated modules now also
give the package “reference”, which is the package-name for upgrade-compatible packages; for other packages it is the package-id.
```